sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -210,6 +210,99 @@ trait GpuProjectExecLike extends ShimUnaryExecNode with GpuExec {
   override def outputBatching: CoalesceGoal = GpuExec.outputBatching(child)
 }
 
+abstract class AbstractProjectSplitIterator(iter: Iterator[ColumnarBatch],
+    schema: Array[DataType],
+    opTime: GpuMetric,
+    numSplitsMetric: GpuMetric) extends Iterator[ColumnarBatch] {
+  private[this] val pending = new scala.collection.mutable.Queue[SpillableColumnarBatch]()
+
+  override def hasNext: Boolean = pending.nonEmpty || iter.hasNext
+
+  def calcNumSplits(cb: ColumnarBatch): Int
+
+  override def next(): ColumnarBatch = {
+    if (!hasNext) {
+      throw new NoSuchElementException()
+    } else if (pending.nonEmpty) {
+      withResource(new MetricRange(opTime)) { _ =>
```

Comment:
Nit: Interesting, didn't realize we had MetricRange.  Seems more complex (object allocation, requires closing) and more verbose than the `ns` method built-in to GpuMetric, i.e.:
```suggestion
      opTime.ns {
```